### PR TITLE
Fixed issues causing errors in test cases using real devices

### PR DESCRIPTION
**Pull request by MounikaBattu17**
**What does this PR accomplish?**
Fixes the errors in tests using real device. The issues are
- Corrected the order of parameters when calling `read_raw` .
- Removed the unwanted parameter `FillMode` passed to function `write_digital_scalar_u32`.
- Updated the ctype of `read_array/write_array` of `read_digital_lines/write_digital_lines` to `bool` since it is expected one.
- Removed using enum `power_up_state` by value in caller since it is returned by value from the interpreter.

**Why should the PR be merged?**
This PR fixes the errors in the testcases mentioned in the [Task 2382630](https://dev.azure.com/ni/DevCentral/_workitems/edit/2382630): Fix failure test cases using hardware

**What testing has been done?**
Ensured errors are not thrown while running these tests but tests are not run using real device.

### CODE REVIEW

File: src/codegen/utilities/interpreter_helpers.py [hunk 1]
```diff
@@ -46,6 +46,8 @@
     "get_analog_power_up_states_with_output_type",
 ]
 
+CTYPE_CHANGE_FUNCTIONS = ["read_digital_lines", "write_digital_lines"]
```
Comment (bkeryan) [inline src/codegen/utilities/interpreter_helpers.py:49]: This is a very generic name for a very specific set of functions. In this case, I think you might as well check against the specific function names in the if-statement or use something like `if "digital_lines" in function.name`.

File: src/codegen/utilities/interpreter_helpers.py [hunk 2]
```diff
@@ -130,7 +132,14 @@ def get_argument_types(functions_metadata):
             if param.is_pointer:
                 argtypes.append(f"ctypes.POINTER({to_param_argtype(param)})")
             else:
-                argtypes.append(to_param_argtype(param))
+                # This is special case. The ctype of these parameters is passed as bool in c call.
+                if (
+                    functions_metadata.function_name in CTYPE_CHANGE_FUNCTIONS
+                    and param.parameter_name in ("read_array", "write_array")
+                ):
+                    argtypes.append("wrapped_ndpointer(dtype=bool, flags=('C'))")
```
Comment (bkeryan) [inline src/codegen/utilities/interpreter_helpers.py:140]: Read and write need different flags: https://numpy.org/doc/stable/reference/routines.ctypeslib.html#numpy.ctypeslib.ndpointer

https://github.com/ni/nidaqmx-python/blob/0.7.0/generated/nidaqmx/_task_modules/read_functions.py
```
wrapped_ndpointer(dtype=bool, flags=('C', 'W')),
```

https://github.com/ni/nidaqmx-python/blob/0.7.0/generated/nidaqmx/_task_modules/write_functions.py#L229
```
wrapped_ndpointer(dtype=bool, flags=('C')),
```

If you move the special case to scrapigen, I think this code will do the right thing.

**Review by bkeryan — CHANGES_REQUESTED** [review]

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: generated/nidaqmx/_library_interpreter.py [hunk 3]
```diff
@@ -4317,7 +4317,7 @@ def read_digital_lines(
                     cfunc.argtypes = [
                         lib_importer.task_handle, ctypes.c_int,
                         ctypes.c_double, ctypes.c_int,
-                        wrapped_ndpointer(dtype=numpy.uint8, flags=('C','W')),
+                        wrapped_ndpointer(dtype=bool, flags=('C')),
                         ctypes.c_uint, ctypes.POINTER(ctypes.c_int),
                         ctypes.POINTER(ctypes.c_int), ctypes.POINTER(c_bool32)]
```
[the remark this answers is not in the document]
Comment (bkeryan) [reply generated/nidaqmx/_library_interpreter.py:4323]: https://github.com/ni/nidaqmx-python/blob/master/src/handwritten/stream_readers.py is passing arguments in the wrong order.

File: src/codegen/utilities/interpreter_helpers.py (hunk 1 above)
Comment (MounikaBattu17) [reply src/codegen/utilities/interpreter_helpers.py]: Added the logic Scarpigen https://github.com/ni/grpc-device-scrapigen/pull/172.

File: src/codegen/utilities/interpreter_helpers.py (hunk 2 above)
Comment (MounikaBattu17) [reply src/codegen/utilities/interpreter_helpers.py]: Moved this logic to Scarpigen. 
I was exploring on these write functions and this PR https://github.com/ni/nidaqmx-python/pull/152 for https://github.com/ni/nidaqmx-python/issues/151 came up, So as expected there, I have made the following changes
1) Differentiated Read and Write functions separately in `functions_addon.py`
2) Passed in Optional parameter `flags="'C','W'"` in `to_param_argtype` and passed `flags` as `"'C'"` for write functions alone
3) Added `is_custom_write_function` method
4) Updated the places accordingly where `CustomCode_Read_Write` is used.

File: src/codegen/utilities/interpreter_helpers.py [hunk 4]
```diff
@@ -130,7 +132,14 @@ def get_argument_types(functions_metadata):
             if param.is_pointer:
                 argtypes.append(f"ctypes.POINTER({to_param_argtype(param)})")
             else:
-                argtypes.append(to_param_argtype(param))
+                # This is special case. The ctype of these parameters is passed as bool in c call.
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply src/codegen/utilities/interpreter_helpers.py]: Moved this to Scrapigen https://github.com/ni/grpc-device-scrapigen/pull/172

File: generated/nidaqmx/_task_modules/in_stream.py [hunk 5]
```diff
@@ -1083,8 +1083,8 @@ def read(self, number_of_samples_per_channel=READ_ALL_AVAILABLE):
         numpy_array = numpy.zeros(number_of_samples, dtype=dtype)
 
         _, samples_read, _ = self._interpreter.read_raw(
-            self._handle, numpy_array, num_samps_per_chan,
-            self.timeout)
+            self._handle, num_samps_per_chan,
+            self.timeout, numpy_array)
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply generated/nidaqmx/_task_modules/in_stream.py:1087]: Sure Brad. The test case for covering `write_raw` method will be handled in separate PR.

File: generated/nidaqmx/_library_interpreter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply generated/nidaqmx/_library_interpreter.py]: Corrected the order.

**Comment by bkeryan** [comment]
Updated results. (This included the direction tweak I suggested, but it probably doesn't matter.)

```
PS D:\dev\nidaqmx-python> poetry run pytest
================================================= test session starts =================================================
platform win32 -- Python 3.9.13, pytest-7.3.1, pluggy-1.0.0
rootdir: D:\dev\nidaqmx-python
plugins: cov-4.0.0, mock-3.10.0
collected 327 items

tests\acceptance\test_examples.py .......s...s..sss.                                                             [  5%]
tests\component\test_buffer_property.py ...                                                                      [  6%]
tests\component\test_channel_properties.py ..................                                                    [ 11%]
tests\component\test_device_properties.py .............                                                          [ 15%]
tests\component\test_export_signal_properties.py .......                                                         [ 18%]
tests\component\test_persisted_channel_properties.py ......                                                      [ 19%]
tests\component\test_persisted_scale_properties.py .........                                                     [ 22%]
tests\component\test_persisted_task_properties.py ......                                                         [ 24%]
tests\component\test_physical_channel_property.py ..........                                                     [ 27%]
tests\component\test_read_properties.py ........................                                                 [ 34%]
tests\component\test_scale.py .                                                                                  [ 35%]
tests\component\test_scale_properties.py .........                                                               [ 37%]
tests\component\test_timing_properties.py ....................                                                   [ 44%]
tests\component\test_trigger_properties.py .............                                                         [ 48%]
tests\component\test_watchdog_properties.py ................                                                     [ 52%]
tests\component\test_write_properties.py ..............                                                          [ 57%]
tests\legacy\test_channel_creation.py .........                                                                  [ 59%]
tests\legacy\test_channels.py ......                                                                             [ 61%]
tests\legacy\test_container_operations.py ...                                                                    [ 62%]
tests\legacy\test_enum_bitfields.py ...                                                                          [ 63%]
tests\legacy\test_events.py .                                                                                    [ 63%]
tests\legacy\test_export_signals.py .                                                                            [ 64%]
tests\legacy\test_invalid_writes.py .....                                                                        [ 65%]
tests\legacy\test_lib.py .                                                                                       [ 66%]
tests\legacy\test_multi_threading.py .                                                                           [ 66%]
tests\legacy\test_power_up_states.py .                                                                           [ 66%]
tests\legacy\test_properties.py .........                                                                        [ 69%]
tests\legacy\test_read_exceptions.py ...                                                                         [ 70%]
tests\legacy\test_read_write.py ...........................                                                      [ 78%]
tests\legacy\test_resource_warnings.py ..                                                                        [ 79%]
tests\legacy\test_stream_analog_readers_writers.py ....                                                          [ 80%]
tests\legacy\test_stream_counter_readers_writers.py ..........                                                   [ 83%]
tests\legacy\test_stream_digital_readers_writers.py ................                                             [ 88%]
tests\legacy\test_stream_power_readers.py ..................                                                     [ 93%]
tests\legacy\test_system_collections.py .....                                                                    [ 95%]
tests\legacy\test_task.py .                                                                                      [ 95%]
tests\legacy\test_teds.py .                                                                                      [ 96%]
tests\legacy\test_triggers.py .....                                                                              [ 97%]
tests\legacy\test_utils.py ....                                                                                  [ 98%]
tests\legacy\test_watchdog.py ..                                                                                 [ 99%]
tests\legacy\test_write_exceptions.py ..                                                                         [100%]

=========================================== 322 passed, 5 skipped in 58.05s ===========================================
```

**Review by bkeryan — CHANGES_REQUESTED** [review]
This looks good, but I think we should use the parameter direction field to decide which flags to use.

File: src/codegen/utilities/interpreter_helpers.py [hunk 6]
```diff
@@ -130,7 +130,10 @@ def get_argument_types(functions_metadata):
             if param.is_pointer:
                 argtypes.append(f"ctypes.POINTER({to_param_argtype(param)})")
             else:
-                argtypes.append(to_param_argtype(param))
+                if is_custom_write_function(functions_metadata):
+                    argtypes.append(to_param_argtype(param, "'C'"))
+                else:
+                    argtypes.append(to_param_argtype(param))
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply src/codegen/utilities/interpreter_helpers.py]: I have updated `flags` to be set from `direction` as per your suggestion. Hope this is fine.

**Comment by MounikaBattu17** [comment]
Thank you Brad for testing the test cases which we couldn't cover.

**Review by bkeryan — APPROVED** [review]

**Pull request merged**